idaes/tests/test_import.py

```diff
@@ -11,81 +11,285 @@
+    if results:
+        for module_name, result in results.items():
+            if result.timing > IMPORT_TIME_MAX:
+                pytest.fail(
+                    f"Import too slow for {module_name}: "
+                    f"{result.timing}s > {IMPORT_TIME_MAX}s"
+                )
+        # success
+        print(f"Time for success: {t1 - t0}s = {(t1 - t0) / BATCH_SIZE}")
+    # If there was a more serious error, look at the batch one-by-one to
+    # figure out which one caused it
+    else:
+        erred = {}
+        for module_name in module_names:
+            result = collect_import_time(
+                module_name, IMPORT_TIME_MAX, allow_output=OUTPUT_OK
+            )
+            if result.error is not None:
+                erred[module_name] = result
+        if erred:  # should always be true
+            errlist = [
+                f"Import error for {module_name}:\n{result.error}"
+                for module_name, result in erred.items()
+            ]
+            pytest.fail("\n".join(errlist))
+
+
+class Result:
+    def __init__(self):
+        self.timing, self.error = -1.0, None
+
+    def add_error(self, e):
+        self.error = e if self.error is None else self.error + "\n" + e
+
+
+def collect_import_time_batch(
+    modules: List[str], max_time: float, allow_output: bool = True
+) -> Union[Dict, None]:
+    """Import a batch of modules by running with 'importtime' option.
+
+    Args:
+        modules: List of absolute name of modules to import (idaes.<something>)
+        max_time: Timeout for process that imports the modules
+        allow_output: If False, consider output during import an error; otherwise
+            ignore it.
+
+    Returns:
+        Either a map of result objects, if all succeeded, or None
+    """
+    results, proc = {m: Result() for m in modules}, None
+
+    # Run import in new interpreter with '-X importtime' option
+    import_stmt = ";".join([f"import idaes.{m}" for m in modules])
+    print(f"import batch: '{import_stmt}'")
+    try:
+        proc = subprocess.run(
+            [sys.executable, "-X", "importtime", "-c", import_stmt],
+            stderr=subprocess.STDOUT,
+            stdout=subprocess.PIPE,
+            timeout=max_time,
+        )
+    except (subprocess.CalledProcessError, subprocess.TimeoutExpired):
+        return None
+
+    if proc.returncode != 0:
+        return None
+
+    # Parse output of '-X importtime'
+    output = proc.stdout.decode()
+    for line in output.splitlines():
+        # For rows with timings
+        if line.startswith("import time:"):
+            # Save if match with a module in the batch
+            for module in modules:
+                if line.endswith(module):
+                    times = re.findall(r"\d+", line)
+                    results[module].timing = float(times[1]) / 1e6
+                    break
+        elif not allow_output:
+            return None
```

Comment:
This is duplicated below in `collect_import_time`; I would recommend pulling out into a function.

Reply:
It's not exactly duplicated, but kinda sorta duplicated. It felt like abstracting might actually be more rather than less confusing. Maybe when @lbianchi-lbl takes out the batching in favor of `.communicate` this becomes moot anyways.